Sources/MongoSwift/SDAM.swift

```diff
@@ -231,29 +254,134 @@ public struct TopologyDescription {
         }
     }
 
+    /// An internal function used to calculate staleness of a given server in a replica set in seconds.
+    internal func staleness(for server: ServerDescription) -> TimeInterval? {
+        guard [.replicaSetWithPrimary, .replicaSetNoPrimary].contains(self.type) else {
+            return nil
+        }
+
+        // From spec: Non-secondary servers have zero staleness
+        guard server.type == .rsSecondary else {
+            return 0.0
+        }
+
+        // Can force unwrap lastWriteDate in here since it's always present in the the ismaster output of primaries
+        // and secondaries, and we filter out all other types of servers.
+        if self.type == .replicaSetWithPrimary {
+            let primary = self.servers.filter { server in server.type == .rsPrimary }[0]
+
+            let delta = { (serverDesc: ServerDescription) -> TimeInterval? in
+                guard let serverLastUpdate = serverDesc.lastUpdateTime else {
+                    return nil
+                }
+                return (serverLastUpdate - serverDesc.lastWriteDate!)
+            }
+
+            guard let clientToServer = delta(server), let clientToPrimary = delta(primary) else {
+                return nil
+            }
+
+            return clientToServer - clientToPrimary + TimeInterval(TopologyDescription.heartbeatFrequencyMS) / 1000.0
+        } else { // ReplicaSetNoPrimary case
+            let secondaries = self.servers.filter { $0.type == .rsSecondary }
+            guard secondaries.count > 0 else {
+                return nil
+            }
+            let sMax = secondaries.max { $0.lastWriteDate! < $1.lastWriteDate! }!
+
+            return (sMax.lastWriteDate! - server.lastWriteDate!)
+                    + TimeInterval(TopologyDescription.heartbeatFrequencyMS) / 1000.0
+        }
+    }
+
+    /// An internal function used to determine if there are suitable servers to read from in a replica set.
+    internal func hasSuitableReadServer(_ readPref: ReadPreference) -> Bool {
+        var suitableModes: [ServerType]
+
+        switch readPref.mode {
+        case .primary:
+            suitableModes = [.rsPrimary]
+        case .secondary:
+            suitableModes = [.rsSecondary]
+        default:
+            suitableModes = [.rsSecondary, .rsPrimary]
+        }
+
+        var candidates = self.servers.filter { suitableModes.contains($0.type) }
+
+        // If read preference specifies a max staleness, first filter out too stale servers
```

Comment:
It does seem like we should be factoring in the latency here. I sort of read it as you did in your second paragraph, but the [read operations](https://github.com/mongodb/specifications/blob/master/source/server-selection/server-selection.rst#read-operations) section of the server selection spec spells it out more explicitly.   

I'll submit another revision that checks latency.

Reply:
Based on https://github.com/mongodb/specifications/blob/master/source/server-selection/server-selection.rst#read-operations, it doesn't look like latency is important. `maxStalenessSeconds` actually factors into narrowing down eligible secondary servers, but `localThresholdMS` is only used to prune multiple eligible servers. Per [Selecting Servers within the Latency Window](https://github.com/mongodb/specifications/blob/master/source/server-selection/server-selection.rst#selecting-servers-within-the-latency-window), the smallest latency among the eligible servers becomes the baseline and you would filter out servers whose latency is too large and finally do a random selection from those that remain. That said, I don't think it will change the return value of `hasReadableServer()` at all so there may be nothing to do for `localThresholdMS`.

Reply:
Oh, I see, the window is defined relative to the available servers. Then yeah, no changes need to be made to accommodate it. That might be why I left it out in the first place but I honestly don't remember.

This has made me realize that I need to change `heartbeatFrequencyMS` from being static to being stored on the `MongoClient`, though, since each client can set its own via the URI. Definitely do not want clients rewriting other clients' settings...